### PR TITLE
[Job submission] Add `list_jobs` API

### LINKED ISSUE #21956: Make a `list_jobs` API
We should export the jobs info to the cluster snapshot as well.

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds an API to the REST server, the SDK, and the CLI for listing all jobs that have been submitted, along with their information.

## Related issue number
RFC: #22412 

Closes #21956 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
test_plasma_unlimited  and test_object_spilling_no_asan  failing on master and unrelated

File: dashboard/modules/job/cli.py [hunk 1]
```diff
@@ -310,3 +311,25 @@ def job_logs(address: Optional[str], job_id: str, follow: bool):
...
+)
+@add_click_logging_options
+def list_jobs(address: Optional[str]):
+    """Lists all running jobs and their information.
+
+    Example:
+        >>> ray job list
+    """
+    client = _get_sdk_client(address)
+    print(pprint.pformat(client.list_jobs()))
```
Comment (edoakes) [inline dashboard/modules/job/cli.py:335]: please use the Ray CLI logger to make sure the output is standardized across calls

**Review by edoakes — COMMENTED** [review]
Looks good, one small comment to use CLI logger.

We'll likely want to add filtering and such in the near future.

**Comment by edoakes** [comment]
@architkulkarni you have merge conflicts

File: dashboard/modules/job/cli.py [hunk 2]
```diff
@@ -314,4 +315,31 @@ def logs(address: Optional[str], job_id: str, follow: bool):
                 "for this feature."
             )
     else:
-        print(client.get_job_logs(job_id), end="")
+        # Set no_format to True because the logs may have unescaped "{" and "}"
+        # and the CLILogger calls str.format().
+        cli_logger.print(client.get_job_logs(job_id), end="", _no_format=True)
```
Comment (edoakes) [inline dashboard/modules/job/cli.py:320]: nit, but if we're using this `_no_format` flag, it should not be `_` prefixed

Comment (edoakes) [reply dashboard/modules/job/cli.py:320]: Changing to cli_logger here doesn't affect the format right?

Comment (edoakes) [reply dashboard/modules/job/cli.py:320]: Or STDOUT vs. STDERR ?

File: dashboard/modules/job/tests/test_http_job_server.py [hunk 3]
```diff
@@ -33,12 +34,61 @@ def headers():
...
-    with _ray_start(include_dashboard=True, num_cpus=1) as address_info:
-        address = address_info["webui_url"]
+    with _ray_start(include_dashboard=True, num_cpus=1) as ctx:
+        address = ctx.address_info["webui_url"]
         assert wait_until_server_available(address)
         yield JobSubmissionClient(format_web_url(address), headers=headers)
 
 
+# NOTE(architkulkarni): This test must be run first in order for the job
+# submission history of the shared Ray runtime to be empty.
```
Comment (edoakes) [inline dashboard/modules/job/tests/test_http_job_server.py:44]: sounds like we need a `DELETE` endpoint...

File: doc/source/cluster/job-submission.rst [hunk 4]
```diff
@@ -113,7 +113,7 @@ Now you may run the following CLI commands:
 
 .. code-block::
 
-    ❯ ray job submit --runtime-env-json='{"working_dir": "./", "pip": ["requests==2.26.0"]}' -- "python script.py"
+    ❯ ray job submit --runtime-env-json='{"working_dir": "./", "pip": ["requests==2.26.0"]}' -- python script.py
```
Comment (edoakes) [inline doc/source/cluster/job-submission.rst:116]: oh good catch...

File: doc/source/cluster/job-submission.rst [hunk 5]
```diff
@@ -300,6 +316,16 @@ Under the hood, both the Job Client and the CLI make HTTP calls to the job serve
...
     logs = rst["logs"]
 
+**List all jobs**
+
+.. code-block:: python
+
+    resp = requests.get(
+        "http://127.0.0.1:8265/api/jobs/"
+    )
+    rst = json.loads(resp.text)
```
Comment (edoakes) [inline doc/source/cluster/job-submission.rst:326]:
```suggestion
    print(resp.json())
```

**Review by edoakes — APPROVED** [review]
Small nits + one clarification, otherwise LGTM

File: dashboard/modules/job/cli.py (hunk 2 above)
Comment (architkulkarni) [reply dashboard/modules/job/cli.py:320]: Should we edit the CLILogger API to change `_no_format` to `no_format`, or skip using the CLILogger altogether?

I'll double check the STDOUT, STDERRand the formatting.

Comment (edoakes) [reply dashboard/modules/job/cli.py:320]: change it to `no_format`

**Comment by edoakes** [comment]
@architkulkarni lint failure

File: dashboard/modules/job/cli.py (hunk 2 above)
Comment (architkulkarni) [reply dashboard/modules/job/cli.py]: Done.  Tested STDOUT and STDERR before and after the change; they were the same.

STDOUT: (colors don't work here but they show up fine in the terminal)
```
[37mJob submission server address[39m: [1mhttp://127.0.0.1:8265[22m

[32m-------------------------------------------------------[39m
[32mJob 'raysubmit_LAQnLBKDs1LuUHEF' submitted successfully[39m
[32m-------------------------------------------------------[39m

[36mNext steps[39m
  Query the logs of the job:
    [1mray job logs raysubmit_LAQnLBKDs1LuUHEF[22m
  Query the status of the job:
    [1mray job status raysubmit_LAQnLBKDs1LuUHEF[22m
  Request the job to be stopped:
    [1mray job stop raysubmit_LAQnLBKDs1LuUHEF[22m

Tailing logs until the job exits (disable with --no-wait):
1
2
3
4
5
6
7
8
9
10
2.26.0

[32m------------------------------------------[39m
[32mJob 'raysubmit_LAQnLBKDs1LuUHEF' succeeded[39m
[32m------------------------------------------[39m

```

STDERR:
```
2022-03-01 15:15:41,728	INFO sdk.py:229 -- Uploading package gcs://_ray_pkg_3dd0f75076634f30.zip.
2022-03-01 15:15:41,729	INFO packaging.py:367 -- Creating a file package for local directory './'.
```

**Pull request merged**